docs/RFCS/structured_configs.md

```diff
@@ -0,0 +1,145 @@
+```SQL
+CREATE TABLE system.users (
+  "username"  CHAR primary key,
+  "hashed"    BLOB
+)
+```
+
+#### Zones
+
+The zone config applies to tables, with a database-level entry used as default.
+When a database is created, it will use a global default.
+When a table is created, it will inherit its database zone config.
+
+Schema:
+```SQL
+CREATE TABLE system.zones (
+  "target"           INT primary key,
+  "attributes"       CHAR,
+  "range_min_bytes"  INT,
+  "range_max_bytes"  INT,
+  "gc_ttl_seconds"   INT
+)
+```
+
+The `target` field is the ID of the database or table the config applies to.
+We will also want a global default which may use the invalid ID 0.
+
+## Privileges
+
+The default privileges for the `system` database and all its tables is `GRANT, SELECT` for
+the `root` only. `root` must always keep these exact permissions. Other users may be
+granted `GRANT, SELECT`, or a subset, but never more.
+
+Since the configurations are mutable, we also allow `INSERT, UPDATE, DELETE` to `root` by
+default and allow these privileges to be granted.
+
+The default `root` privileges and maximum grantable privileges now become:
+
+| Table name | Maximum privileges                    |
+|------------|---------------------------------------|
+| namespace  | GRANT, SELECT                         |
+| descriptor | GRANT, SELECT                         |
+| accounting | GRANT, SELECT, INSERT, UPDATE, DELETE | 
+| users      | GRANT, SELECT, INSERT, UPDATE, DELETE | 
+| zones      | GRANT, SELECT, INSERT, UPDATE, DELETE | 
+
+## Setting configuration parameters
+
+The table structure and privilege settings allow modification of the configuration
+through sql commands.
+
+For ease of use and bulk modifications, the current command-line commands
+should be modified to function with the new format.
+
+# Implementation
+
+Due to the current state of the system, this RFC will implement table-based configuration
+in multiple steps:
+
+| Configuration | Implementation date |
+|---------------|---------------------|
+| accounting    | immediately         |
+| permissions   | after [#2089](https://github.com/cockroachdb/cockroach/issues/2089) |
+| users         | immediately         |
+| zones         | after [#2179](https://github.com/cockroachdb/cockroach/issues/2179) |
+
+# Drawbacks
+
+# Alternatives
+
+Some of the configurations could be stored in the `descriptor` table.
+However, we would need to implement column-based privileges to allow modification of the configuration
+while preventing any changes to the descriptors.
+
+The specific column names and types for each table are flexible. Specifically, the accounting
+configuration contents are unknown.
+
+# Unresolved questions
+
+It may be desirable to move the privilege configuration from the database and table descriptors and
+store them in their own table.
+
+The database/table IDs in the config tables are not particularly clear. It would be nice to
+dynamically show/use the database/table names.
```

Comment:
If we implement a `SHOW ZONES` statement, then this could be done using a JOIN on the `descriptor` table. (We'd also have to be able to pull the name out of the descriptor or change the descriptor table to store the name separately).